src/libopensc/card-epass2003.c

```diff
@@ -456,36 +461,50 @@ aes128_encrypt_cmac(const unsigned char *key, int keysize,
 }
 
 static int
-aes128_encrypt_ecb(const unsigned char *key, int keysize,
+aes128_encrypt_ecb(struct sc_card *card, const unsigned char *key, int keysize,
 		const unsigned char *input, size_t length, unsigned char *output)
 {
 	unsigned char iv[EVP_MAX_IV_LENGTH] = { 0 };
-	return openssl_enc(EVP_aes_128_ecb(), key, iv, input, length, output);
+	EVP_CIPHER *alg = sc_evp_cipher(card->ctx, "AES-128-ECB");
+	int r;
+	r = openssl_enc(alg, key, iv, input, length, output);
+	sc_evp_cipher_free(alg);
+	return r;
 }
 
 
 static int
-aes128_encrypt_cbc(const unsigned char *key, int keysize, unsigned char iv[16],
+aes128_encrypt_cbc(struct sc_card *card, const unsigned char *key, int keysize, unsigned char iv[16],
 		const unsigned char *input, size_t length, unsigned char *output)
 {
-	return openssl_enc(EVP_aes_128_cbc(), key, iv, input, length, output);
+	EVP_CIPHER *alg = sc_evp_cipher(card->ctx, "AES-128-CBC");
+	int r;
+	r = openssl_enc(alg, key, iv, input, length, output);
+	sc_evp_cipher_free(alg);
+	return r;
 }
 
 
 static int
-aes128_decrypt_cbc(const unsigned char *key, int keysize, unsigned char iv[16],
+aes128_decrypt_cbc(struct sc_card *card, const unsigned char *key, int keysize, unsigned char iv[16],
 		const unsigned char *input, size_t length, unsigned char *output)
 {
-	return openssl_dec(EVP_aes_128_cbc(), key, iv, input, length, output);
+	EVP_CIPHER *alg = sc_evp_cipher(card->ctx, "AES-128-CBC");
```

Comment:
do we need some error checking in this case or can the openssl functions handle NULL `EVP_CIPHER` in reasonable way (like returning an error instead of crashing).

Reply:
Not really, the next function will fail if we pass in a NULL cipher anyway, so I thought that is sufficient, these are all hard-coded names, so it is extremely unlikely a failure would happen anyway.

Reply:
(It is no more likely to fail than the previous version where a function was called inlined in the Init() calls.